internal/lang/evalchecks/eval_for_each.go

```diff
@@ -145,82 +152,102 @@ func performForEachValueChecks(expr hcl.Expression, hclCtx *hcl.EvalContext, all
-		return cty.UnknownVal(cty.Map(cty.DynamicPseudoType)), diags
+		return cty.NullVal(cty.Map(cty.DynamicPseudoType)), diags
 	case markSafeLengthInt(forEachVal) == 0:
 		// If the map is empty ({}), return an empty map, because cty will
 		// return nil when representing {} AsValueMap. This also covers an empty
 		// set (toset([]))
 		return forEachVal, diags
+	case ty.IsSetType():
+		setVal, setTypeDiags := performSetTypeChecks(expr, hclCtx, allowUnknown, allowedTypesMessage, forEachVal, excludableAddr)
+		diags = diags.Append(setTypeDiags)
+		if diags.HasErrors() {
+			return setVal, diags
+		}
+	}
+
+	return forEachVal, diags
+}
+
+func performForEachValueChecksUnknown(expr hcl.Expression, hclCtx *hcl.EvalContext, allowUnknown bool, allowedTypesMessage string, forEachVal cty.Value, excludableAddr addrs.Targetable) (cty.Value, tfdiags.Diagnostics) {
+	var diags tfdiags.Diagnostics
+	ty := forEachVal.Type()
+
+	if !allowUnknown {
+		var detailMsg string
+		if ty.IsSetType() {
+			detailMsg = errInvalidUnknownDetailSet
+		} else {
+			detailMsg = errInvalidUnknownDetailMap
+		}
+		detailMsg += forEachCommandLineExcludeSuggestion(excludableAddr)
+
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity:    hcl.DiagError,
+			Summary:     "Invalid for_each argument",
+			Detail:      detailMsg,
+			Subject:     expr.Range().Ptr(),
+			Expression:  expr,
+			EvalContext: hclCtx,
+			Extra:       DiagnosticCausedByUnknown(true),
+		})
+
+		return cty.UnknownVal(cty.Map(cty.DynamicPseudoType)), diags
 	}
 
 	if ty.IsSetType() {
-		setVal, setTypeDiags := performSetTypeChecks(expr, hclCtx, allowUnknown, forEachVal, excludableAddr)
+		setVal, setTypeDiags := performSetTypeChecks(expr, hclCtx, allowUnknown, allowedTypesMessage, forEachVal, excludableAddr)
 		diags = diags.Append(setTypeDiags)
 		if diags.HasErrors() {
 			return setVal, diags
 		}
+		return cty.UnknownVal(cty.Set(cty.DynamicPseudoType)), diags
 	}
 
-	return forEachVal, diags
+	// ensure that we have a map, and not a DynamicValue
+	return cty.UnknownVal(cty.Map(cty.DynamicPseudoType)), diags
 }
 
 // performSetTypeChecks does checks when we have a Set type, as sets have some gotchas
-func performSetTypeChecks(expr hcl.Expression, hclCtx *hcl.EvalContext, allowUnknown bool, forEachVal cty.Value, excludableAddr addrs.Targetable) (cty.Value, tfdiags.Diagnostics) {
+func performSetTypeChecks(expr hcl.Expression, hclCtx *hcl.EvalContext, allowUnknown bool, allowedTypesMessage string, forEachVal cty.Value, excludableAddr addrs.Targetable) (cty.Value, tfdiags.Diagnostics) {
 	var diags tfdiags.Diagnostics
 	ty := forEachVal.Type()
 
+	if ty.ElementType() != cty.String && ty.ElementType() != cty.DynamicPseudoType {
+		var extra interface{}
+		if !forEachVal.IsWhollyKnown() {
+			extra = DiagnosticCausedByUnknown(true)
+		}
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity:    hcl.DiagError,
+			Summary:     "Invalid for_each set argument",
```

Comment:
Can you explain to me why this diagnostic is telling the user about allowed types, when we know they already passed a set with a bad element type?